Author: Qingtao Cao <qingtao.cao@gmail.com>
Date:   Tue Feb 26 09:56:13 2019 +1000

    OG-6686 udhcpc: bind to the downstream interface when sending unicast packets
    
    When sending unicast packets such as to renew or to release the current lease,
    bind to the downstream interface that udhcpc has been specified to cater for
    
    This helps ensure unicast packets are routed via the specified interface even
    if its default route doesn't have the highest priority
    
    BTW, previously only in LISTEN_KERNEL mode, would the udhcpc bound udp sockets
    to the interface just to receive unicast replies from server side.
Index: busybox-1.31.0/networking/udhcp/packet.c
===================================================================
--- busybox-1.31.0.orig/networking/udhcp/packet.c
+++ busybox-1.31.0/networking/udhcp/packet.c
@@ -189,7 +189,7 @@ int FAST_FUNC udhcp_send_raw_packet(stru
 /* Let the kernel do all the work for packet generation */
 int FAST_FUNC udhcp_send_kernel_packet(struct dhcp_packet *dhcp_pkt,
 		uint32_t source_nip, int source_port,
-		uint32_t dest_nip, int dest_port)
+		uint32_t dest_nip, int dest_port, const char *interface)
 {
 	struct sockaddr_in sa;
 	unsigned padding;
@@ -204,6 +204,13 @@ int FAST_FUNC udhcp_send_kernel_packet(s
 	}
 	setsockopt_reuseaddr(fd);
 
+	if (interface) {
+		if (setsockopt_bindtodevice(fd, interface) == -1) {
+			msg = "setsockopt(%s)";
+			goto ret_close;
+		}
+	}
+
 	memset(&sa, 0, sizeof(sa));
 	sa.sin_family = AF_INET;
 	sa.sin_port = htons(source_port);
Index: busybox-1.31.0/networking/udhcp/dhcpc.c
===================================================================
--- busybox-1.31.0.orig/networking/udhcp/dhcpc.c
+++ busybox-1.31.0/networking/udhcp/dhcpc.c
@@ -773,7 +773,7 @@ static int bcast_or_ucast(struct dhcp_pa
 	if (server)
 		return udhcp_send_kernel_packet(packet,
 			ciaddr, CLIENT_PORT,
-			server, SERVER_PORT);
+			server, SERVER_PORT, client_data.interface);
 	return raw_bcast_from_client_data_ifindex(packet, ciaddr);
 }
 
Index: busybox-1.31.0/networking/udhcp/dhcpd.c
===================================================================
--- busybox-1.31.0.orig/networking/udhcp/dhcpd.c
+++ busybox-1.31.0/networking/udhcp/dhcpd.c
@@ -604,7 +604,7 @@ static void send_packet_to_relay(struct
 
 	udhcp_send_kernel_packet(dhcp_pkt,
 			server_data.server_nip, SERVER_PORT,
-			dhcp_pkt->gateway_nip, SERVER_PORT);
+			dhcp_pkt->gateway_nip, SERVER_PORT, NULL);
 }
 
 static void send_packet(struct dhcp_packet *dhcp_pkt, int force_broadcast)
Index: busybox-1.31.0/networking/udhcp/common.h
===================================================================
--- busybox-1.31.0.orig/networking/udhcp/common.h
+++ busybox-1.31.0/networking/udhcp/common.h
@@ -318,7 +318,7 @@ int udhcp_send_raw_packet(struct dhcp_pa
 
 int udhcp_send_kernel_packet(struct dhcp_packet *dhcp_pkt,
 		uint32_t source_nip, int source_port,
-		uint32_t dest_nip, int dest_port) FAST_FUNC;
+		uint32_t dest_nip, int dest_port, const char *interface) FAST_FUNC;
 
 void udhcp_sp_setup(void) FAST_FUNC;
 void udhcp_sp_fd_set(struct pollfd *pfds, int extra_fd) FAST_FUNC;
